File: Dia12/d12p2/src/main.py
def path(node, connections, visited, double):
	if node == "end":
		return [["end"]]
	res = []
	visited = visited.copy()
	visited.append(node)
	for c in connections[node]:
		if c.isupper() or c not in visited:
			# print(node, ">", c)
			for x in path(c, connections, visited, double):
				x.insert(0, node)
				res.append(x)
	if double:
		for c in connections[node]:
			if not c.isupper() and c in visited and c != "start":
				# print(node, ">", c)
				for x in path(c, connections, visited, False):
					x.insert(0, node)
					res.append(x)
	return res



with open('../../input.txt') as f:
	connections = {}
	for [n0, n1] in map(lambda x: x.strip().split('-'), f.readlines()):
		if n0 not in connections:
			connections[n0] = []
		if n1 not in connections:
			connections[n1] = []
		connections[n0].append(n1)
		connections[n1].append(n0)
	# print(connections["start"])
	print(len(path("start", connections, [], True)))

File: Dia12/d12p2/src/main.js
function path(node, connections, visited, double) {
	if (node === "end") {
		return [["end"]]
	}
	let res = []
	visited = [node, ...visited]
	for (let c of connections[node]) {
		if (!visited.includes(c) || c === c.toUpperCase()) {
			for (let x of path(c, connections, visited, double)) {
				res.push([node, ...x])
			}
		}else if(double && c !== "start") {
			for (let x of path(c, connections, visited, false)) {
				res.push([node, ...x])
			}
		}
	}
	// console.log(res.length, res[0])
	return res
}

readFileSync = require('fs').readFileSync;
connections = {}
for(let line of readFileSync('../../input.txt', 'utf8').split('\n').map(v => v.trim().split('-'))) {
	if (connections[line[0]] === undefined) {
		connections[line[0]] = []
	}
	if (connections[line[1]] === undefined) {
		connections[line[1]] = []
	}
	connections[line[0]].push(line[1])
	connections[line[1]].push(line[0])
}

console.log(path("start", connections, [], true).length)